src/dbnode/storage/limits/query_limits.go

```diff
@@ -0,0 +1,243 @@
+func NewLookbackLimit(name string, scope tally.Scope, opts LookbackLimitOptions) LookbackLimit {
+	return &lookbackLimit{
+		name:    name,
+		options: opts,
+		metrics: lookbackLimitMetrics{
+			recent:     scope.Gauge(fmt.Sprintf("recent-%s", name)),
+			recentPeak: scope.Gauge(fmt.Sprintf("recent-peak-%s", name)),
+			total:      scope.Counter(fmt.Sprintf("total-%s", name)),
+			exceeded:   scope.Tagged(map[string]string{"limit": name}).Counter("limit-exceeded"),
+		},
+		recent: atomic.NewInt64(0),
+		stopCh: make(chan struct{}),
+	}
+}
+
+func (q *queryLimits) DocsLimit() LookbackLimit {
+	return q.docsLimit
+}
+
+func (q *queryLimits) BytesReadLimit() LookbackLimit {
+	return q.bytesReadLimit
+}
+
+func (q *queryLimits) Start() {
+	q.docsLimit.Start()
+	q.bytesReadLimit.Start()
+}
+
+func (q *queryLimits) Stop() {
+	q.docsLimit.Stop()
+	q.bytesReadLimit.Stop()
+}
+
+func (q *queryLimits) AnyExceeded() error {
+	if err := q.docsLimit.Exceeded(); err != nil {
+		return err
+	}
+	if err := q.bytesReadLimit.Exceeded(); err != nil {
+		return err
+	}
+	return nil
+}
+
+// Inc increments the current value and returns an error if above the limit.
+func (q *lookbackLimit) Inc(new int) error {
+	if q == nil {
+		return nil
+	}
+	if new < 0 {
+		return fmt.Errorf("invalid negative query limit inc %d", new)
+	}
+	if new == 0 {
+		return q.Exceeded()
+	}
+
+	// Add the new stats to the global state.
+	newI64 := int64(new)
+	recent := q.recent.Add(newI64)
+
+	// Update metrics.
+	q.metrics.recent.Update(float64(recent))
+	q.metrics.total.Inc(newI64)
+
+	// Enforce limit (if specified).
+	return q.exceeded(recent)
+}
+
+func (q *lookbackLimit) Exceeded() error {
+	return q.exceeded(q.recent.Load())
+}
+
+func (q *lookbackLimit) exceeded(recent int64) error {
+	if q.options.Limit > 0 && recent > q.options.Limit {
+		q.metrics.exceeded.Inc(1)
+		return fmt.Errorf(
+			"query aborted due to limit: name=%s, limit=%d, current=%d, within=%s",
+			q.name, q.options.Limit, recent, q.options.Lookback)
+	}
+	return nil
+}
+
+// Start initializes background processing for enforcing the query limit.
+func (q *lookbackLimit) Start() {
+	if q == nil {
+		return
+	}
+	ticker := time.NewTicker(q.options.Lookback)
+	go func() {
+		for {
+			select {
+			case <-ticker.C:
+				q.reset()
+			case <-q.stopCh:
+				ticker.Stop()
+				return
+			}
+		}
+	}()
+}
+
+func (q *lookbackLimit) Stop() {
+	if q == nil {
+		return
+	}
+	close(q.stopCh)
+}
+
+// Validate returns an error if the query stats options are invalid.
+func (opts LookbackLimitOptions) Validate() error {
+	if opts.Limit < 0 {
+		return fmt.Errorf("query limit requires limit >= 0 (%d)", opts.Limit)
+	}
+	if opts.Lookback <= 0 {
+		return fmt.Errorf("query limit requires lookback > 0 (%d)", opts.Lookback)
+	}
+	return nil
+}
+
+func (q *lookbackLimit) current() int64 {
+	return q.recent.Load()
+}
+
+func (q *lookbackLimit) reset() {
+	// Update peak gauge only on resets so it only tracks
```

Comment:
it took me a bit to understand the logic here. i wonder if better names could help. you're essentially doing a standard prev, curr swap right? Maybe `prevPeak` and `currPeak`?

Reply:
hmm maybe not since you are actually emitting 2 metrics and `currPeak` isn't actually a peak. why do you need a `recent` metric in addition to the `recentPeak` metric? seems like `recentPeak` is what you care about for determining the proper limit for a lookback period.

Reply:
Yeah I agree the distinction isn't very obvious. Maybe `recent_count` and `recent_max` since the former is the raw count and the latter is the max/peak per period?

But so in general my thinking is it is just better to include both the original and newer implementation (instead of just newer) since then we always will have either at our disposal and we can actually compare the two (it would be unfortunate if we regret changing the implementation after rolling out to prod).